File: src/main/java/com/github/sorusclient/client/module/impl/oldanimations/v1_8_9/OldAnimationsTransformer.java
package com.github.sorusclient.client.module.impl.oldanimations.v1_8_9;

import com.github.glassmc.loader.GlassLoader;
import com.github.glassmc.loader.Listener;
import com.github.glassmc.loader.util.Identifier;
import com.github.sorusclient.client.transform.Transformer;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.*;

public class OldAnimationsTransformer extends Transformer implements Listener {

    @Override
    public void run() {
        GlassLoader.getInstance().registerTransformer(this.getClass());
    }

    public OldAnimationsTransformer() {
        this.register("v1_8_9/net/minecraft/client/render/item/HeldItemRenderer", this::transformHeldItemRenderer);
        this.register("v1_8_9/net/minecraft/client/MinecraftClient", this::transformMinecraftClient);
        this.register("v1_8_9/net/minecraft/client/render/entity/feature/ArmorFeatureRenderer", this::transformArmorFeatureRenderer);
    }

    private void transformHeldItemRenderer(ClassNode classNode) {
        Identifier method_1354 = Identifier.parse("v1_8_9/net/minecraft/client/render/item/HeldItemRenderer#method_1354(F)V");
        Identifier method_9873 = Identifier.parse("v1_8_9/net/minecraft/client/render/item/HeldItemRenderer#method_9873(FF)V");

        for (MethodNode methodNode : classNode.methods) {
            if (methodNode.name.equals(method_1354.getMethodName()) && methodNode.desc.equals(method_1354.getMethodDesc())) {
                for (AbstractInsnNode node : methodNode.instructions) {
                    if (node instanceof MethodInsnNode && ((MethodInsnNode) node).name.equals(method_9873.getMethodName()) && ((MethodInsnNode) node).desc.equals(method_9873.getMethodDesc()) && node.getPrevious().getOpcode() == Opcodes.T_LONG) {
                        methodNode.instructions.remove(node.getPrevious());

                        InsnList insnList = new InsnList();
                        insnList.add(new VarInsnNode(Opcodes.FLOAD, 4));
                        insnList.add(new MethodInsnNode(Opcodes.INVOKESTATIC, OldAnimationsHook.class.getName().replace(".", "/"), "getPartialTicks", "(F)F"));
                        methodNode.instructions.insertBefore(node, insnList);
                    }
                }
            }
        }
    }

    private void transformMinecraftClient(ClassNode classNode) {
        Identifier tick = Identifier.parse("v1_8_9/net/minecraft/client/MinecraftClient#tick()V");

        for (MethodNode methodNode : classNode.methods) {
            if (methodNode.name.equals(tick.getMethodName()) && methodNode.desc.equals(tick.getMethodDesc())) {
                methodNode.instructions.insert(new MethodInsnNode(Opcodes.INVOKESTATIC, OldAnimationsHook.class.getName().replace(".", "/"), "updateSwing", "()V"));
            }
        }
    }

    private void transformArmorFeatureRenderer(ClassNode classNode) {
        Identifier combineTextures = Identifier.parse("v1_8_9/net/minecraft/client/render/entity/feature/ArmorFeatureRenderer#combineTextures()Z");

        for (MethodNode methodNode : classNode.methods) {
            if (methodNode.name.equals(combineTextures.getMethodName()) && methodNode.desc.equals(combineTextures.getMethodDesc())) {
                InsnList insnList = new InsnList();
                insnList.add(new MethodInsnNode(Opcodes.INVOKESTATIC, OldAnimationsHook.class.getName().replace(".", "/"), "showArmorDamage", "()Z"));
                insnList.add(new InsnNode(Opcodes.IRETURN));

                methodNode.instructions.insert(insnList);
            }
        }
    }

}

File: version/1.8.9/src/main/kotlin/com/github/sorusclient/client/feature/impl/oldanimations/v1_8_9/OldAnimationsTransformer.kt
package com.github.sorusclient.client.feature.impl.oldanimations.v1_8_9

import com.github.sorusclient.client.toIdentifier
import com.github.sorusclient.client.transform.Transformer
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.*

@Suppress("UNUSED")
class OldAnimationsTransformer : Transformer() {

    init {
        register("v1_8_9/net/minecraft/client/render/item/HeldItemRenderer", this::transformHeldItemRenderer)
        register("v1_8_9/net/minecraft/client/MinecraftClient", this::transformMinecraftClient)
        register("v1_8_9/net/minecraft/client/render/entity/feature/ArmorFeatureRenderer", this::transformArmorFeatureRenderer)
    }

    private fun transformHeldItemRenderer(classNode: ClassNode) {
        val method1354 = "v1_8_9/net/minecraft/client/render/item/HeldItemRenderer#method_1354(F)V".toIdentifier()
        val method9873 = "v1_8_9/net/minecraft/client/render/item/HeldItemRenderer#method_9873(FF)V".toIdentifier()
        for (methodNode in classNode.methods) {
            if (methodNode.name == method1354.methodName && methodNode.desc == method1354.methodDesc) {
                for (node in methodNode.instructions) {
                    if (node is MethodInsnNode && node.name == method9873.methodName && node.desc == method9873.methodDesc && node.getPrevious().opcode == Opcodes.T_LONG) {
                        methodNode.instructions.remove(node.getPrevious())
                        val insnList = InsnList()
                        insnList.add(VarInsnNode(Opcodes.FLOAD, 4))
                        insnList.add(
                            MethodInsnNode(
                                Opcodes.INVOKESTATIC,
                                OldAnimationsHook::class.java.name.replace(".", "/"),
                                "getPartialTicks",
                                "(F)F"
                            )
                        )
                        methodNode.instructions.insertBefore(node, insnList)
                    }
                }
            }
        }
    }

    private fun transformMinecraftClient(classNode: ClassNode) {
        val tick = "v1_8_9/net/minecraft/client/MinecraftClient#tick()V".toIdentifier()
        for (methodNode in classNode.methods) {
            if (methodNode.name == tick.methodName && methodNode.desc == tick.methodDesc) {
                methodNode.instructions.insert(
                    MethodInsnNode(
                        Opcodes.INVOKESTATIC,
                        OldAnimationsHook::class.java.name.replace(".", "/"),
                        "updateSwing",
                        "()V"
                    )
                )
            }
        }
    }

    private fun transformArmorFeatureRenderer(classNode: ClassNode) {
        val combineTextures =
            "v1_8_9/net/minecraft/client/render/entity/feature/ArmorFeatureRenderer#combineTextures()Z".toIdentifier()
        for (methodNode in classNode.methods) {
            if (methodNode.name == combineTextures.methodName && methodNode.desc == combineTextures.methodDesc) {
                val insnList = InsnList()
                insnList.add(
                    MethodInsnNode(
                        Opcodes.INVOKESTATIC,
                        OldAnimationsHook::class.java.name.replace(".", "/"),
                        "showArmorDamage",
                        "()Z"
                    )
                )
                insnList.add(InsnNode(Opcodes.IRETURN))
                methodNode.instructions.insert(insnList)
            }
        }
    }

}